From c0e232dc3cf22557b82b0f985642a660188bb754 Mon Sep 17 00:00:00 2001
From: Robert Kanter <rkanter@apache.org>
Date: Wed, 4 May 2016 22:42:21 -0700
Subject: [PATCH 1501/2848] MAPREDUCE-6675. TestJobImpl.testUnusableNode
 failed (haibochen via rkanter)

(cherry picked from commit 9d3fcdfbb314c83ba6185e4ac8de649dad51a279)

Change-Id: I1b1a43d3b5e7d2514cca2ecb524f9e4d23e6ff71
---
 .../mapreduce/v2/app/job/impl/TestJobImpl.java     |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestJobImpl.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestJobImpl.java
index 2af4380..f126f3d 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestJobImpl.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestJobImpl.java
@@ -529,7 +529,7 @@ public void testUnusableNodeTransition() throws Exception {
     Configuration conf = new Configuration();
     conf.set(MRJobConfig.MR_AM_STAGING_DIR, stagingDir);
     conf.setInt(MRJobConfig.NUM_REDUCES, 1);
-    AsyncDispatcher dispatcher = new AsyncDispatcher();
+    DrainDispatcher dispatcher = new DrainDispatcher();
     dispatcher.init(conf);
     dispatcher.start();
     CyclicBarrier syncBarrier = new CyclicBarrier(2);
@@ -606,6 +606,7 @@ public void handle(TaskAttemptEvent event) {
     NodeReport secondMapperNodeReport = nodeReports.get(1);
     job.handle(new JobUpdatedNodesEvent(job.getID(),
         Collections.singletonList(firstMapperNodeReport)));
+    dispatcher.await();
     // complete the reducer
     for (TaskId taskId: job.tasks.keySet()) {
       if (taskId.getTaskType() == TaskType.REDUCE) {
-- 
1.7.9.5

